# CIRI3

## About

CIRI3 is a comprehensive analysis package for the detection and quantification of circRNAs in RNA-Seq data, while providing differential expression analysis of circRNAs at multiple levels.

## Installation

The CIRI3 was constructed based on java. The `environment.yaml` was provided and the dependencies can be installed as the follow:
```
git clone https://github.com/gyjames/CIRI3.git
cd CIRI3
conda env create -n CIRI3 -f ./environment.yaml
conda activate CIRI3
```

## Usage

### Step1. Mapping

#### 1)BWA
Recommended protocols for running BWA-MEM:
```
bwa index -a bwtsw ref.fa
bwa mem –T 19 ref.fa reads.fq > sample.sam (single-end reads)
bwa mem –T 19 ref.fa read1.fq read2.fq > sample.sam (paired-end reads)
```
* ref.fa: FASTA file of all reference sequences

#### 2)STAR
```
STAR --runThreadN 10 \
--genomeDir star_index \
--outSAMtype SAM \
--readFilesIn reads1.fq read2.fq \
--outFileNamePrefix output_dir/ \
--outReadsUnmapped Fastx \
--outSJfilterOverhangMin 15 12 12 12 \
--alignSJoverhangMin 15 \
--alignSJDBoverhangMin 15 \
--outFilterMultimapNmax 20 \
--outFilterScoreMin 1 \
--outFilterMatchNmin 1 \
--outFilterMismatchNmax 2 \
--chimSegmentMin 15 \
--chimScoreMin 15 \
--chimJunctionOverhangMin 15

cd output_dir/
bwa mem -T 19 ref.fa Unmapped.out.mate1 Unmapped.out.mate2 > bwa.sam
```

### Step2. CircRNA detection and quantification

CIRI3 provides multiple input options for the identification and quantification of circRNAs, including single-sample input, multiple-sample input, multiple-sample files containing RNase R treated information. In addition, users have the option to input a collection of circRNAs of interest, allowing CIRI3 to quantify the BSJ and FSJ of these circRNAs within the samples.

A small test data set is provided in [data/](data/). The unpacked files are:

#### 1)Single-SAM/BAM files as input
CIRI3 will give information about circRNAs in the sample.
```
# without the annotation gtf
java -jar CIRI3.jar -I ./data/circRNA/Single/sample.sam -O ./data/circRNA/Single/result.txt -F ./data/circRNA/ref.fa (BWA-MEM)
java -jar CIRI3.jar -I /path/to/Chimeric.out.junction,/path/to/Aligned.out.sam,/path/to/bwa.sam -O output.ciri3 -F ref.fa -Ma 1 (STAR)

# with the annotation gtf
Java -jar CIRI3.jar -I sample.sam -O result.txt -F ref.fa -A ref.gtf (BWA-MEM)
java -jar CIRI3.jar -I /path/to/Chimeric.out.junction,/path/to/Aligned.out.sam,/path/to/bwa.sam -O output.ciri3 -F ref.fa -A ref.gtf -Ma 1 (STAR)

```
* sample.sam and bwa.sam: SAM file generated by BWA-MEM
* Aligned.out.sam and Chimeric.out.junction: output files generated by STAR
* ref.gtf: GTF/GFF3 formatted annotation file

#### 2)Multiple-SAM/BAM files as input
CIRI3 will give information about circRNA in the samples, circRNA BSJ expression matrix and circRNA FSJ expression matrix.
```
# without the annotation gtf
java -jar CIRI3.jar -I ./data/circRNA/Mutiple/samples.tsv -O ./data/circRNA/Mutiple/result.txt -F ./data/circRNA/ref.fa -W 1 (BWA-MEM)
java -jar CIRI3.jar -I samples.tsv -O output.ciri3 -F ref.fa -Ma 1 -W 1 (STAR)

# with the annotation gtf
Java -jar CIRI3.jar -I samples.tsv -O result.txt -F ref.fa -A ref.gtf -W 1 (BWA-MEM)
java -jar CIRI3.jar -I samples.tsv -O output.ciri3 -F ref.fa -A ref.gtf -Ma 1 -W 1 (STAR)
```
* samples.tsv is a tab separated file like (BWA-MEM):
  + Column 1: absolute path to the sam/bam file
```
/path/to/sample1.sam
/path/to/sample2.sam
/path/to/sample3.sam
/path/to/sample4.sam
```
* samples.tsv is a tab separated file like (STAR):
  + Column 1: absolute path to the sam/bam file
```
/path/to/sample1/Chimeric.out.junction,/path/to/sample1/Aligned.out.sam,/path/to/sample1/bwa.sam
/path/to/sample2/Chimeric.out.junction,/path/to/sample2/Aligned.out.sam,/path/to/sample2/bwa.sam
/path/to/sample3/Chimeric.out.junction,/path/to/sample3/Aligned.out.sam,/path/to/sample3/bwa.sam
/path/to/sample4/Chimeric.out.junction,/path/to/sample4/Aligned.out.sam,/path/to/sample4/bwa.sam
```
#### 3)Multiple-SAM/BAM files containing RNase R treated information as input
CIRI3 gives information on circRNAs in the samples, circRNA BSJ expression matrix, circRNA FSJ expression matrix and circRNA enrichment ratios.
```
# without the annotation gtf
java -jar CIRI3.jar -I ./data/circRNA/RNase/sample_infor.tsv -O ./data/circRNA/RNase/result.txt -F ./data/circRNA/ref.fa -W 2

# with the annotation gtf
Java -jar CIRI3.jar -I sample_infor.tsv -O result.txt -F ref.fa -A ref.gtf -W 2
```
* samples.tsv is a tab separated file like:
  + Column 1: absolute path to the sam/bam file
  + Column 2: RNase R treated information. 0 (without RNase R treated)    1 (with RNase R treated)   2 (RNase R treatment unknown)
```
/path/to/sample1.sam  0
/path/to/sample2.sam  0
/path/to/sample3.sam  1
/path/to/sample4.sam  1
```

#### 4)CircRNA collections of interest and sam/bam files as inputs
CIRI3 will give the BSJ expression matrix and FSJ expression matrix of the input circRNA collection in the sample.
```
java -jar CIRI3.jar -I ./data/circRNA/User/sample.sam -UC ./data/circRNA/User/circRNA_List.txt -O ./data/circRNA/User/result.txt -F ./data/circRNA/ref.fa
```

* circRNA_List.txt is a tab separated file like:
  + #Column 1: chromosome
  + #Column 2: start loci of a circRNA on the chromosome(1-based)
  + #Column 3: end loci of a circRNA on the chromosome
```
chr1	1379083	1414681
chr1	1384101	1397537
```

### Step3. Differential Expression Analysis

CIRI3 provides three levels of differential expression analysis algorithms, including the expression of a single circRNA, the junction ratio of a single circRNA, and the relative expression of multiple circRNAs derived from a single gene. The entire differential expression analysis script incorporates parts of the code from the edgeR package in R or rMATS to calculate the statistical significance of differential expression (p-values).

#### 1)the expression of a single circRNA

##### Study without biological replicate(Input in two ways)

```
java -jar CIRI3.jar DE_BSJ -I ./data/DE/BSJ/Without_RE/infor.tsv -O ./data/DE/BSJ/Without_RE/result.txt

java -jar CIRI3.jar DE_BSJ -I ./data/DE/BSJ/Without_RE/infor.tsv -M ./data/DE/BSJ/Without_RE/BSJ_Matrix.txt -O ./data/DE/BSJ/Without_RE/result.txt
```

* infor.tsv is a tab separated file like:
  + Column 1: sample name
  + Column 2: absolute path to CIRI3 output result files
  + Column 3: sample type
  + Column 4: number of mapped reads (It can be obtained from the log file output by CIRI3)
```
Sample	Path	Class	Map_reads
Case1	path/to/case1_CIRI3.txt	Case	1050326
Control1	path/to/control1_CIRI3.txt	Control	1055952
```

* BSJ_Matrix.txt is a tab separated file like:
```
circRNA_ID	Case1	Control1
chr9:33886881|33900271	5	6
chr9:83311887|83343287	4	0
chr9:33318731|33338589	0	6
```

##### Study with biological replicate(Input in two ways)

```
java -jar CIRI3.jar DE_BSJ -I ./data/DE/BSJ/With_RE/infor.tsv -G ./data/DE/BSJ/With_RE/Gene_Expression.txt -O ./data/DE/BSJ/With_RE/result.txt

java -jar CIRI3.jar DE_BSJ -I ./data/DE/BSJ/With_RE/infor.tsv -G ./data/DE/BSJ/With_RE/Gene_Expression.txt -M ./data/DE/BSJ/With_RE/BSJ_Matrix.txt -O ./data/DE/BSJ/With_RE/result.txt
```

* infor.tsv is a tab separated file like:
  + Column 1: sample name
  + Column 2: absolute path to CIRI3 output result files
  + Column 3: sample type
  + Column 4: subject (optional, only for paired samples)
```
Sample	Path	Class	Num
Case1	path/to/case1_CIRI3.txt	Case	1
Case2	path/to/case2_CIRI3.txt	Case	2
Control1	path/to/control1_CIRI3.txt	Control	1
Control2	path/to/control2_CIRI3.txt	Control	2
```

* Gene_Expression.txt is a tab separated file like Gene_Expression.txt is a tab separated file like (which can be obtained by running featureCounts on a BWA-aligned and sorted BAM files):
```
Geneid	Case1	Case2	Control1	Control2
ENSG00000236875.3	14	4	13	4
ENSG00000181404.18	59	21	37	10
```

#### 2)the junction ratio of a single circRNA

```
java -jar CIRI3.jar DE_Ratio -I ./data/DE/Ratio/infor.tsv -BM ./data/DE/Ratio/BSJ_Matrix.txt -FM ./data/DE/Ratio/FSJ_Matrix.txt -O ./data/DE/Ratio/result.txt
```

* infor.tsv is a tab separated file like:
  + Column 1: sample name
  + Column 2: sample type
```
Sample	Class
Case1	Case
Case2	Case
Control1	Control
Control2	Control
```

* FSJ_Matrix.txt is a tab separated file like:
```
circRNA_ID	Case1	Case2	Control1	Control2
chr5:171214|173067	15	0	12	0
chr1:18854213|18857631	5	6	0	0
```

#### 3)the relative expression of multiple circRNAs derived from a single gene(Input in two ways)

```
java -jar CIRI3.jar DE_Relative -I ./data/DE/Relative_Expression/infor.tsv -O ./data/DE/Relative_Expression/result.txt

java -jar CIRI3.jar DE_Relative -I ./data/DE/Relative_Expression/infor.tsv -M ./data/DE/Relative_Expression/BSJ_Matrix.txt -GC ./data/DE/Relative_Expression/circ_Gene.txt -O ./data/DE/Relative_Expression/result.txt
```

* infor.tsv is a tab separated file like:
  + Column 1: sample name
  + Column 2: absolute path to CIRI3 output result files
  + Column 3: sample type
```
Sample	Path	Class
Case1	path/to/case1_CIRI3.txt	Case
Case2	path/to/case2_CIRI3.txt	Case
Control1	path/to/control1_CIRI3.txt	Control
Control2	path/to/control2_CIRI3.txt	Control
```

* circ_Gene is a tab separated file like:
```
circRNA_ID	gene_id
chr9:465962|467219	ENSG00000227155.7
chr9:846960|894195	ENSG00000137090.12
chr9:2161686|2181676	ENSG00000080503.24
```

### All Arguments

```
java -jar CIRI3.jar -H

Arguments:

    -I, --in
          Input SAM file name or SAM files list(required; generated by BWA-MEM)
    -O, --out
          Output circRNA file name(required)
    -F, --ref_file
          FASTA file of all reference sequences. Please make sure this file is
          the same one provided to BWA-MEM (required).
    -A, --anno
          input GTF/GFF3 formatted annotation file name (optional)
    -G, --log
          output log file name (optional)
    -H, --help
          show this help information
    -Max, --max_span
          max spanning distance of circRNAs (default: 200000)
    -Min, --min_span
          min spanning distance of circRNAs (default: 140)
    -S, --strigency
          2: only output circRNAs supported by more than 2 distinct PCC signals (default)
          1: only output circRNAs supported by more than 2 junction reads
          0: output all circRNAs regardless junction read or PCC signal counts
    -U, --mapq_uni
          set threshold for mappqing quality of each segment of junction reads
          (default: 10; should be within [0,30])
    -E, --rel_exp
          set threshold for relative expression calculated based on counts of
          junction reads and non-junction reads (optional: e.g. 0.1)
    -Mc, --mitochondria
          0: Skip the recognition of  mitochondria circRNA (default)
          1: Perform the recognition of mitochondria circRNA (the ID of mitochondrion in reference file is required)
    -M, --chrM
          tell CIRI3 the ID of mitochondrion in reference file(s) (default:
          chrM)
    -T, --thread_num
          set number of threads for parallel running (default: 1)
    -It, --intron
          0: Skip the recognition of  intronic self-ligated circRNA (default)
          1: Perform the recognition of intronic self-ligated circRNA (formatted annotation file is required)
    -Sp, --splicing_signals
	    0: Only canonical GT-AG splice signals were considered (default)
	    1: Both canonical GT-AG and non-canonical splice signals were considered
    -Ma, --mapper\r\n"
	    0: The SAM file generated by the BWA-MEM (default)
	    1: the SAM file generated by STAR
    -W, --way
          0: Input a single SAM file to identify circRNA (default)
          1: Input a SAM files list (including the absolute path to SAM files
             and sample name) to identify circRNA
          2: Input a SAM files list (including the absolute path to SAM files,
             RNase information, and sample names) to identify circRNA, this can generate
             circRNA confidence score
    -UC, --user_circ
          file of circRNA collections of interest to users

```

```
java -jar CIRI3.jar DE_BSJ -H

Arguments:

    -I, --in
          the file of sample list(required)
    -O, --out
          output differential expression result(required)
    -M, --matrix
          circRNA BSJ expression matrix (optional)
    -G, --gene
          gene expression matrix
    -P, --pvalue
          p value threshold for DE score calculation (default: 0.05)
    -H, --help
          show this help information
```

```
java -jar CIRI3.jar DE_Ratio -H

Arguments:

    -I, --in
          the file of sample list(required)
    -O, --out
          output differential expression result(required)
    -BM, --bsj_matrix
          circRNA BSJ expression matrix (required)
    -FM, --fsj_matrix
          circRNA FSJ expression matrix (required)
    -T, --thread_num
          set number of threads for parallel running (default: 1)
    -H, --help
          show this help information
```

```
java -jar CIRI3.jar DE_Relative -H

Arguments:

    -I, --in
          the file of sample list(required)
    -O, --out
          output differential expression result(required)
    -M, --matrix
          circRNA BSJ expression matrix (optional)
    -GC, --circ_gene
          the gene information file corresponding to circRNAs (optional)
    -T, --thread_num
          set number of threads for parallel running (default: 1)
    -H, --help
          show this help information
```

## Output

The three main output files are:
* `result.txt` is an information file for predicted circRNA.
  + Column 1: ID of a predicted circRNA in the pattern of "chr:start|end";
  + Column 2: chromosome of a predicted circRNA
  + Column 3: start loci of a predicted circRNA on the chromosome
  + Column 4: end loci of a predicted circRNA on the chromosome
  + Column 5: circular junction read (also called as back-spliced junction read) count of a predicted circRNA 
  + Column 6: unique CIGAR types of a predicted circRNA. For example, a circRNAs have three junction reads: read A (80M20S, 80S20M), read B (80M20S, 80S20M), read 
              C (40M60S, 40S30M30S, 70S30M), then its has two SM types (80S20M, 70S30M), two MS types (80M20S, 70M30S) and one SMS type (40S30M30S). Thus its 
              SM_MS_SMS should be 2_2_1.
  + Column 7: non-junction read count of a predicted circRNA that mapped across the circular junction but consistent with linear RNA instead of being back-spliced
  + Column 8: ratio of circular junction reads calculated by 2*#junction_reads/(2*#junction_reads+#non_junction_reads). 
  + Column 9: type of a circRNA according to positions of its two ends on chromosome (exon, intron or intergenic_region; only available when annotation file is 
              provided)
  + Column 10: ID of the gene(s) where an exonic or intronic circRNA locates
  + Column 11: strand info of a predicted circRNAs (new in CIRI2)
  + Column 12: all of the circular junction read IDs (split by ",")
  + Column 13: the count of high-quality BSJ reads from the first scan
* `result.txt.BSJ_Matrix` is a BSJ expression matrix file for predicted circRNA.
  + Each detected isoform is reported on a separate line.
* `result.txt.BSJ_Matrix` is a FSJ expression matrix file for predicted circRNA.
* `result.txt.Score` is a enrichment ratio information file for predicted circRNA.




## References

1. Li, Heng. "Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM." arXiv preprint arXiv:1303.3997 (2013).
2. Shen S, Park JW, Lu ZX, et al. rMATS: robust and flexible detection of differential alternative splicing from replicate RNA-Seq data. Proc Natl Acad Sci U S A. 2014;111(51):E5593-E5601. doi:10.1073/pnas.1419161111
3. Robinson MD, McCarthy DJ, Smyth GK. edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics. 2010;26(1):139-140. doi:10.1093/bioinformatics/btp616
4. Liao Y, Smyth GK, Shi W. featureCounts: an efficient general purpose program for assigning sequence reads to genomic features. Bioinformatics. 2014;30(7):923-930. doi:10.1093/bioinformatics/btt656